File: BCI2000Native_StimulusPresentation/StimulusPresentationScript_DemoShort.py
import sys
import os
import random

default = 'C:\\bci2000.x64\\tools\\python\\'
BCI2000_python_tools = input('Enter path to BCI2000 python tools (default is '+default+'): ')
if not BCI2000_python_tools:
   BCI2000_python_tools = default

sys.path.append(BCI2000_python_tools)

from convert_bciprm import *

def float_range( start, stop, step ):
    l = []
    i = start
    q = 0
    while i < stop:
        l.append( i )
        q += 1
        i = start + q * step
    return l

print( "----Loading Settings----" )

settings = BCISettings() 

settings[ "BCI2000Path" ] = os.path.join( BCI2000_python_tools, "..", ".." )
print( "BCI2000Path: ", settings.BCI2000Path )


settings[ "TaskDuration         " ] = '2s'
settings[ "ISIDuration          " ] = '1.5s'
settings[ "InstructionDuration  " ] = '30s'
settings[ "CaptionSwitch        " ] = '1'
settings[ "parm_filename        " ] = os.path.join( settings[ "BCI2000Path" ], "parms", "demo_parms_py.prm" )

settings.InstructCaption   = [ 'Stimulus Presentation Task. Press space to continue', 'End of task.' ]


print( "----Done Loading Settings----" )

print( "----Reading Default Params----" )

param = read_bciprm( os.path.join( BCI2000_python_tools, "default.prm" ) )

print( "----Done Reading Params----" )

print( "----Setting Params----" )

n_stimuli = 62
n_rows = 7

param[ 'Application' ]['Stimuli']['Stimuli']['Section        '] = 'Application'
param[ 'Application' ]['Stimuli']['Stimuli']['Type           '] = 'matrix'
param[ 'Application' ]['Stimuli']['Stimuli']['DefaultValue   '] = ''
param[ 'Application' ]['Stimuli']['Stimuli']['LowRange       '] = ''
param[ 'Application' ]['Stimuli']['Stimuli']['HighRange      '] = ''
param[ 'Application' ]['Stimuli']['Stimuli']['Comment        '] = 'captions and icons to be displayed, sounds to be played for different stimuli'
param[ 'Application' ]['Stimuli']['Stimuli']['Value          '] = bcimatrix( n_rows, n_stimuli )
param[ 'Application' ]['Stimuli']['Stimuli']['RowLabels      '] = bcilist( n_rows )
param[ 'Application' ]['Stimuli']['Stimuli']['ColumnLabels   '] = bcilist( n_stimuli )

param.Application.Stimuli.Stimuli.RowLabels[0]  = 'caption'
param.Application.Stimuli.Stimuli.RowLabels[1]  = 'icon'
param.Application.Stimuli.Stimuli.RowLabels[2]  = 'audio'
param.Application.Stimuli.Stimuli.RowLabels[3]  = 'StimulusDuration'
param.Application.Stimuli.Stimuli.RowLabels[4]  = 'AudioVolume'
param.Application.Stimuli.Stimuli.RowLabels[5]  = 'Category'
param.Application.Stimuli.Stimuli.RowLabels[6]  = 'EarlyOffsetExpression'

directory   = os.path.join( settings[ "BCI2000Path" ], "prog", "images" )
task_images = os.listdir( directory )
task_images = [f for f in task_images if os.path.isfile(directory+'/'+f)]

for i in task_images:
    if not i.endswith('.bmp'):
        task_images.remove(i)

random.shuffle( task_images )

# Study images 1-50
for idx in range( 0, len( task_images ) ):
    param.Application.Stimuli.Stimuli.ColumnLabels[ idx ]    = str( idx + 1 )
    param.Application.Stimuli.Stimuli.Value[ 0 ][ idx ]      = ''
    param.Application.Stimuli.Stimuli.Value[ 1 ][ idx ]      = os.path.join( "..", "prog", "images", task_images[ idx ] )
    param.Application.Stimuli.Stimuli.Value[ 2 ][ idx ]      = ''
    param.Application.Stimuli.Stimuli.Value[ 3 ][ idx ]      = settings.TaskDuration
    param.Application.Stimuli.Stimuli.Value[ 4 ][ idx ]      = '0'      
    param.Application.Stimuli.Stimuli.Value[ 5 ][ idx ]      = 'image' 
    param.Application.Stimuli.Stimuli.Value[ 6 ][ idx ]      = '' 


# inter-stimulus interval (fixation cross) 51
idx=50
param.Application.Stimuli.Stimuli.ColumnLabels[ idx ]    = str( idx + 1 )
param.Application.Stimuli.Stimuli.Value[ 0 ][ idx ]      = '+'
param.Application.Stimuli.Stimuli.Value[ 1 ][ idx ]      = ''
param.Application.Stimuli.Stimuli.Value[ 2 ][ idx ]      = ''
param.Application.Stimuli.Stimuli.Value[ 3 ][ idx ]      = str( settings.ISIDuration )
param.Application.Stimuli.Stimuli.Value[ 4 ][ idx ]      = '0' 
param.Application.Stimuli.Stimuli.Value[ 5 ][ idx ]      = 'fixation'
param.Application.Stimuli.Stimuli.Value[ 6 ][ idx ]      = ''


# Instructions 61-62
idx_iter = 0
for idx in range( 60, 60 + len( settings.InstructCaption ) ):
    param.Application.Stimuli.Stimuli.ColumnLabels[ idx ] = str( idx )
    param.Application.Stimuli.Stimuli.Value[ 0 ][ idx ]      = settings.InstructCaption[ idx_iter ]
    param.Application.Stimuli.Stimuli.Value[ 1 ][ idx ]      = ''
    param.Application.Stimuli.Stimuli.Value[ 2 ][ idx ]      = ''
    param.Application.Stimuli.Stimuli.Value[ 3 ][ idx ]      = settings.InstructionDuration
    param.Application.Stimuli.Stimuli.Value[ 4 ][ idx ]      = '0'    
    param.Application.Stimuli.Stimuli.Value[ 5 ][ idx ]      = 'instruction' 
    param.Application.Stimuli.Stimuli.Value[ 6 ][ idx ]      = 'KeyDown == 32' # space key 
    
    idx_iter = idx_iter + 1

# Sequence
taskseq = []
for img in range( 1, len(task_images) + 1 ):
    taskseq += [ 51, img ]

seq = [ 61, *taskseq, 62 ]

param.Application.Sequencing.Sequence.Section      = 'Application'
param.Application.Sequencing.Sequence.Type         = 'intlist'
param.Application.Sequencing.Sequence.DefaultValue = '1'
param.Application.Sequencing.Sequence.LowRange     = '1'
param.Application.Sequencing.Sequence.HighRange    = ''
param.Application.Sequencing.Sequence.Comment      = 'Stimuli sequence (deterministic mode)'
param.Application.Sequencing.Sequence.Value        = [ str( i ) for i in seq ]
param.Application.Sequencing.Sequence.NumericValue = seq 

#
param.Application.Sequencing.SequenceType.Section              = 'Application'
param.Application.Sequencing.SequenceType.Type                 = 'int'
param.Application.Sequencing.SequenceType.DefaultValue         = '0'
param.Application.Sequencing.SequenceType.LowRange             = '0'
param.Application.Sequencing.SequenceType.HighRange            = '1'
param.Application.Sequencing.SequenceType.Comment              = 'Sequence of stimuli is 0 deterministic, 1 random (enumeration)'
param.Application.Sequencing.SequenceType.Value                = '0'

#
param.Application.Stimuli.CaptionSwitch.Section       = 'Application'
param.Application.Stimuli.CaptionSwitch.Type          = 'int'
param.Application.Stimuli.CaptionSwitch.DefaultValue  = '1'
param.Application.Stimuli.CaptionSwitch.LowRange      = '0'
param.Application.Stimuli.CaptionSwitch.HighRange     = '1'
param.Application.Stimuli.CaptionSwitch.Comment       = 'Present captions (boolean)'
param.Application.Stimuli.CaptionSwitch.Value         = [ settings.CaptionSwitch ]

#
param.Application.Stimuli.CaptionHeight.Section      = 'Application'
param.Application.Stimuli.CaptionHeight.Type         = 'int'
param.Application.Stimuli.CaptionHeight.DefaultValue = '0'
param.Application.Stimuli.CaptionHeight.LowRange     = '0'
param.Application.Stimuli.CaptionHeight.HighRange    = '100'
param.Application.Stimuli.CaptionHeight.Comment      = 'Height of stimulus caption text in percent of screen height'
param.Application.Stimuli.CaptionHeight.Value        = [ '5' ]

print( "----Done Setting Params----" )
print( "----Converting Params----" )

convert_bciprm( param, settings.parm_filename )

print( f"----Params Saved to { settings.parm_filename }----" )

File: BCI2000Native_StimulusPresentation/StimulusPresentationScript_DemoShort.m
% BCI2000 Stimulus Presentation Demo Script
% 
% StimulusPresentationScript_Demo creates a parameter fragment that can be
% loaded into BCI2000 to create a stimulus presentation experiment.
% 
% This demo script will take the image files located in the BCI2000 prog
% directory and create a stimuli matrix containing these images, variable
% duration fixation cross stimuli, instructions, and a sync pulse. 
% 
% Change the n_rows and n_stimuli variables to store more information with
% the stimuli or add additional stimuli. Best practice is to separate
% stimuli into banks (e.g. 1-25, 101-125, etc) for easy evaluation later. 
% 
% Note that every stimulus needs to have an index for every row desired,
% even if that row label is not meaningful for the stimulus.
% 
% A sequence is created to alternate the fixation cross stimuli with the
% image stimuli.
% 
% The stimuli and meaningful parameters are written into a param
% variable and stored as a *.prm file using the convert_bciprm function.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Author: James Swift <swift@neurotechcenter.org>
%%
%% $BEGIN_BCI2000_LICENSE$
%% 
%% This file is part of BCI2000, a platform for real-time bio-signal research.
%% [ Copyright (C) 2000-2021: BCI2000 team and many external contributors ]
%% 
%% BCI2000 is free software: you can redistribute it and/or modify it under the
%% terms of the GNU General Public License as published by the Free Software
%% Foundation, either version 3 of the License, or (at your option) any later
%% version.
%% 
%% BCI2000 is distributed in the hope that it will be useful, but
%%                         WITHOUT ANY WARRANTY
%% - without even the implied warranty of MERCHANTABILITY or FITNESS FOR
%% A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
%% 
%% You should have received a copy of the GNU General Public License along with
%% this program.  If not, see <http://www.gnu.org/licenses/>.
%% 
%% $END_BCI2000_LICENSE$
%% http://www.bci2000.org 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Set the path of the BCI2000 main directory here
prompt               = {'Enter path to BCI2000 root directory: '};
dlgtitle             = 'Enter BCI2000 path';
fieldsize            = [1 45];
definput             = {fullfile('C:','bci2000.x64')};
userinput            = inputdlg(prompt,dlgtitle,fieldsize,definput);
settings.BCI2000path = userinput{1};

% Add BCI2000 tools to path
addpath(genpath(fullfile(settings.BCI2000path,'tools')))

%% Settings
settings.TaskDuration        = '2s';
settings.ISIDuration         = '1.5s';
settings.InstructionDuration = '30s';

settings.parm_filename       = fullfile(settings.BCI2000path,'parms','demo_parms.prm');

settings.InstructCaption     = {'Stimulus Presentation Task. Press space to continue'; 'End of task.'};

%% Get task images
task_images = dir(fullfile(settings.BCI2000path,'prog','images','*.bmp'));

%% Set up unique stimulus codes, separated into banks for easy evaluation
n_stimuli = 62; % Total events
n_rows    = 7;

% break down into blocks for easier analysis later
% 1-50:  image stimuli
% 51:    inter-stimulus interval (variable duration)
% 61-62: instructions

% Set up Stimuli
param.Stimuli.Section         = 'Application';
param.Stimuli.Type            = 'matrix';
param.Stimuli.DefaultValue    = '';
param.Stimuli.LowRange        = '';
param.Stimuli.HighRange       = '';
param.Stimuli.Comment         = 'captions and icons to be displayed';
param.Stimuli.Value           = cell(n_rows,n_stimuli);
param.Stimuli.Value(:)        = {''};
param.Stimuli.RowLabels       = cell(n_rows,1);
param.Stimuli.RowLabels(:)    = {''};
param.Stimuli.ColumnLabels    = cell(1,n_stimuli);
param.Stimuli.ColumnLabels(:) = {''};

param.Stimuli.RowLabels{1}  = 'caption';
param.Stimuli.RowLabels{2}  = 'icon';
param.Stimuli.RowLabels{3}  = 'audio';
param.Stimuli.RowLabels{4}  = 'StimulusDuration';
param.Stimuli.RowLabels{5}  = 'AudioVolume';
param.Stimuli.RowLabels{6}  = 'Category';
param.Stimuli.RowLabels{7}  = 'EarlyOffsetExpression';

%% Study images 1-50
for idx = 1:length(task_images)
    param.Stimuli.ColumnLabels{idx} = sprintf('%d',idx);
    param.Stimuli.Value{1,idx}      = '';
    param.Stimuli.Value{2,idx}      = sprintf('%s',...
                fullfile('..','prog','images',task_images(idx).name));
    param.Stimuli.Value{3,idx}      = '';
    param.Stimuli.Value{4,idx}      = settings.TaskDuration;
    param.Stimuli.Value{5,idx}      = '0';      
    param.Stimuli.Value{6,idx}      = 'image'; 
    param.Stimuli.Value{7,idx}      = ''; 
end 


%% inter-stimulus interval (fixation cross) 51
idx = 51;

param.Stimuli.ColumnLabels{idx} = sprintf('%d',idx);
param.Stimuli.Value{1,idx}      = '+';
param.Stimuli.Value{2,idx}      = '';
param.Stimuli.Value{3,idx}      = '';
param.Stimuli.Value{4,idx}      = num2str(settings.ISIDuration,7);
param.Stimuli.Value{5,idx}      = '0';      
param.Stimuli.Value{6,idx}      = 'fixation'; 
param.Stimuli.Value{7,idx}      = ''; 


%% Instructions 61-62
idx_iter = 1;
for idx = 61:60+length(settings.InstructCaption)
    param.Stimuli.ColumnLabels{idx} = sprintf('%d',idx);
    param.Stimuli.Value{1,idx}      = settings.InstructCaption{idx_iter};
    param.Stimuli.Value{2,idx}      = '';
    param.Stimuli.Value{3,idx}      = '';
    param.Stimuli.Value{4,idx}      = settings.InstructionDuration;
    param.Stimuli.Value{5,idx}      = '0';    
    param.Stimuli.Value{6,idx}      = 'instruction'; 
    param.Stimuli.Value{7,idx}      = 'KeyDown == 32'; % space key 
    
    idx_iter = idx_iter + 1;
end     


%% Sequence
% 1-50:    image stimuli
% 51:      inter-stimulus interval (variable duration)
% 61-62: instructions

randOrder = randperm(length(task_images));
taskseq   = [];
for i = 1:length(task_images)
    currentImage = randOrder(i);
    taskseq      = [taskseq 51 currentImage];
end

seq = [ 61 taskseq 62 ]';


param.Sequence.Section      = 'Application';
param.Sequence.Type         = 'intlist';
param.Sequence.DefaultValue = '1';
param.Sequence.LowRange     = '1';
param.Sequence.HighRange    = '';
param.Sequence.Comment      = 'Sequence in which stimuli are presented';
param.Sequence.Value        = cellfun(@num2str, num2cell(seq), 'un',0);
param.Sequence.NumericValue = seq;


%% Sequence Type
param.SequenceType.Section      = 'Application';
param.SequenceType.Type         = 'int';
param.SequenceType.DefaultValue = '0';
param.SequenceType.LowRange     = '0';
param.SequenceType.HighRange    = '1';
param.SequenceType.Comment      = '0 deterministic, 1 random (enumeration)';
param.SequenceType.Value        = {'0'};


%% Caption Switch
param.CaptionSwitch.Section      = 'Application';
param.CaptionSwitch.Type         = 'int';
param.CaptionSwitch.DefaultValue = '1';
param.CaptionSwitch.LowRange     = '0';
param.CaptionSwitch.HighRange    = '1';
param.CaptionSwitch.Comment      = 'Present captions (boolean)';
param.CaptionSwitch.Value        = {'1'};


%% write the param struct to a bci2000 parameter file
parameter_lines = convert_bciprm( param );
fid = fopen(settings.parm_filename, 'w');

for i=1:length(parameter_lines)
    fprintf( fid, '%s', parameter_lines{i} );
    fprintf( fid, '\r\n' );
end
fclose(fid);
